File: LTI1.3_Tools/F社/サンプルコード/app/Http/Controllers/LtiController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use Auth;
use App\Models\User;
use Exception;
use App\Http\Shares\LtiDatabase;
use IMSGlobal\LTI;

class LtiController extends Controller
{
    /**
     * Open Id Connect Login Request
     *
     * @param
     *    Request request   リクエストパラーメタ
     *
     * @return なし（リダイレクト）
     */
    public function login(Request $request) {
        try {
            // ログイン初期化要求を受け取り、プラットフォームに戻る
            LTI\LTI_OIDC_Login::new(new LtiDatabase())
            ->do_oidc_login_redirect($request['target_link_uri'])
            ->do_redirect();
        // ログイン初期化要求でエラーが発生した場合は画面上にエラーメッセージを出力
        } catch(Exception $e) {
            echo $e->getMessage();
        }
    }

    /**
     * LTI Message Launches
     *
     * @param
     *    Request request   リクエストパラーメタ
     *
     * @return なし（リダイレクト）
     */
    public function resourceLink(Request $request) {
        try {
            // メッセージを起動し、各種バリデーションチェックを行う
            $launch = LTI\LTI_Message_Launch::new(new LtiDatabase())
            ->validate();
        // 各種バリデーションチェックでエラーが発生した場合は画面上にエラーメッセージを出力
        } catch(Exception $e) {
            echo $e->getMessage();
            return;
        }

        // subを取得
        $user_account_id = $launch->get_launch_data()['sub'] ?? null;
        Log::debug('###LtiController::resourceLink##id_token###', ['fields' => $request['id_token']]);
        Log::debug('###LtiController::resourceLink##state###', ['fields' => $request['state']]);
        Log::debug('###LtiController::resourceLink##sub###', ['fields' => $user_account_id]);

        // iatが空の場合はエラー
        if (empty($launch->get_launch_data()['iat'])) {
            echo 'iat is empty';
            return;
        }

        // expが空の場合はエラー
        if (empty($launch->get_launch_data()['exp'])) {
            echo 'exp is empty';
            return;
        }

        // rolesを取得
        $roles = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/roles'] ?? null;
        // rolesが空の場合はエラー
        if (empty($roles)) {
            echo 'roles is empty';
            return;
        }
        $role_array = [
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Administrator",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Alumni",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Faculty",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Guest",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Learner",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Member",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Mentor",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#None",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Observer",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Other",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#ProspectiveStudent",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Staff",
            "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Student",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Administrator",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#ContentDeveloper",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Learner",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Manager",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Member",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Mentor",
            "http://purl.imsglobal.org/vocab/lis/v2/membership#Officer",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#AccountAdmin",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#Administrator",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#Creator",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#None",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#SysAdmin",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#SysSupport",
            "http://purl.imsglobal.org/vocab/lis/v2/system/person#User"
        ];
        foreach($roles as $role) {
            // roleが指定値以外の場合はエラー
            if (!in_array($role, $role_array)) {
                echo 'Invalid roles【' . $role . '】';
                return;
            }
        }

        // target_link_uriが空の場合はエラー
        if (empty($launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/target_link_uri'])) {
            echo 'target_link_uri is empty';
            return;
        }

        // contextを取得
        $context = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/context'] ?? null;
        // contextのidを取得
        $context_id = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/context']['id'] ?? null;
        // contextが存在し、contextのidが空の場合はエラー
        if (!empty($context) && empty($context_id)) {
            echo 'context_id is empty';
            return;
        }

        // gradeを取得
        $custom_grade = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/custom']['grade'] ?? null;
        $custom_grade_array = ['P1', 'P2', 'P3', 'P4', 'P5', 'P6', 'J1', 'J2', 'J3'];
        // customのgradeが存在し、指定値以外の場合はエラー
        if (!empty($custom_grade) && !in_array($custom_grade, $custom_grade_array)) {
            echo 'Invalid custom_grade【' . $custom_grade . '】';
            return;
        }

        $UUIDv4 = '/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i';

        // tool_platformを取得
        $tool_platform = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/tool_platform'] ?? null;
        // tool_platformのguidを取得
        $tool_platform_gu_id = $launch->get_launch_data()['https://purl.imsglobal.org/spec/lti/claim/tool_platform']['guid'] ?? null;
        // tool_platformが存在し、tool_platformのgu_idが空の場合はエラー
        if (!empty($tool_platform) && empty($tool_platform_gu_id)) {
            echo 'tool_platform_guid is empty';
            return;
        // tool_platformのgu_idが空以外でUUIDv4形式以外の場合はエラー
        } else if (!empty($tool_platform_gu_id) && !preg_match($UUIDv4, $tool_platform_gu_id)) {
            echo 'Invalid tool_platform_guid【' . $tool_platform_gu_id . '】';
            return;
        }

        // subが空以外でUUIDv4形式以外の場合はエラー
        if (!empty($user_account_id) && !preg_match($UUIDv4, $user_account_id)) {
            echo 'Invalid sub【' . $user_account_id . '】';
            return;
        // subが空、または、subがユーザー情報に存在しない場合はトップ画面を表示
        } else if (empty($user_account_id) || User::where('account_id', $user_account_id)->get()->isEmpty()) {
            Log::debug('###LtiController::resourceLink##not exist account_id###');
            return view('top');
        } else {
            // ユーザー情報を取得してログイン
            $user = User::where('account_id', $user_account_id)->first();    
            Auth::login($user);

            // roleが1（生徒）の場合は、生徒のページにリダイレクト
            if ($user->role == 1) {
                Log::debug('###LtiController::resourceLink##exist student account_id###');
                return redirect('/students/lessons');
            // roleが2（先生）の場合は、先生のページにリダイレクト
            } elseif ($user->role == 2) {
                Log::debug('###LtiController::resourceLink##exist teacher account_id###');
                return redirect('/teachers/lessons');
            }
        }
    }
}

File: LTI1.3_Tools/N社/サンプルコード/app/controllers/api/student/lti_controller.rb
module Api
  module Student
    # 生徒がLTIを通じてログインするためのController
    class LtiController < ApplicationController
      # 弊社独自領域ここから---
      ## 各メソッド実行前に認証チェックを行わない
      skip_before_action :my_authenticate_user
      # ---弊社独自領域ここまで

      # ResourceLinkRequestにより実行されるメソッド
      # Callback URLへPOST疑似リダイレクトする
      def implicit_back
        domain = request.headers["SERVER_NAME"]

        render html: Repost::Senpai.perform(
          "https://#{domain}:#{request.port}/api/auth/student_lti/callback",
          params: { utf8: params[:utf8], id_token: params[:id_token], state: params[:state], commit: params[:commit] }
        ).html_safe
      end

      # Callback phase後に実行されるメソッド
      # omniauth_openid_connectではチェックできていないパラメータの入力チェックを行い、エラー出力、またはログインする
      def launches
        # パラメータの入力チェックを行う
        valid_lti_params

        # 弊社独自領域ここから---
        ## "sub"で生徒をデータベースから検索する
        student_info = if auth_hash['uid'].present?
                         StudentInfo.joins(:student)
                                    .where(student: {
                                             uuid: auth_hash['uid'],
                                             id_status: %i[active locked]
                                           })
                                    .order(nendo: :desc).first
                       end
        ## 生徒が見つからない場合、エラーをレスポンスする
        if student_info.blank?
          render_err 'UUIDが一致する児童生徒が見つかりません。'
          return
        end

        ## 生徒が所属する学校をデータベースから取得する
        school = student_info.school

        ## 学校毎に分かれているFQDNを取得する
        to_domain = if school.student_domain
                      "#{school.student_domain}#{Rails.application.config.app_name[:student_domain]}"
                    end
        ## FQDNが無い場合、エラーをレスポンスする
        if to_domain.blank?
          render_err '学校のドメインが設定されていません。'
          return
        end
        ## アクセスしてきているURLのFQDNと学校のFQDNが違う場合、学校のFQDNに置換してPOST疑似リダイレクトする
        domain = request.headers["SERVER_NAME"]
        if domain != to_domain
          render html: Repost::Senpai.perform(
            "https://#{to_domain}:#{request.port}/api/auth/student_lti/callback",
            params: { utf8: params[:utf8], id_token: params[:id_token], state: params[:state], commit: params[:commit] }
          ).html_safe
          return
        end

        logger.info 'Sucessfully logged'
        logger.info "UUID: #{auth_hash['uid']}"
        ## 生徒アカウントのロック状態を確認する(認証失敗が規定回数を超える、または手動変更でアカウントがロックできる仕様)
        error = SessionController.check_lock student_info
        if error.present?
          ## ロックされている場合、エラーをレスポンスする
          render_err "ログインに失敗しました。\nError: #{error}"
        else
          ## そうでない場合、ログイン処理を行い、ログイン直後の画面へリダイレクトする
          SessionController.before_sign_in student_info, request
          bypass_sign_in student_info.student
          student_info.student.need_password_change = false
          student_info.student.save!
          redirect_to '/#/student'
        end
        # ---弊社独自領域ここまで
      end

      # omniauth_openid_connectによる処理(Request phase・Callback phase)中にエラーが発生した際にリダイレクトされることにより実行されるメソッド
      # エラーをレスポンスする
      def failure
        render_err "ログインに失敗しました。\nError: #{params[:message]}"
      end

      private

      # プラットフォームから返ってきた情報を取得する
      def auth_hash
        request.env['omniauth.auth']
      end

      CLAIM = 'https://purl.imsglobal.org/spec/lti/claim/'.freeze # 末尾スラッシュ有
      # omniauth_openid_connectではチェックできていないパラメータの入力チェックを行う
      def valid_lti_params
        # "sub"が空か無効値の場合、例外を発生する
        uid = auth_hash['uid']
        unless uid.present? && valid_uuid?(uid)
          raise InvalidLtiParam, { param_key: 'sub', param_value: uid }
        end

        # "iat"が空か0の場合、例外を発生する
        iat = auth_hash.extra.raw_info['iat']
        raise InvalidLtiParam, { param_key: 'iat', param_value: iat } if iat.blank? || iat.zero?

        # "message_type"が「LtiResourceLinkRequest」でない場合、例外を発生する
        message_type = auth_hash.extra.raw_info["#{CLAIM}message_type"]
        unless message_type == 'LtiResourceLinkRequest'
          raise InvalidLtiParam, { param_key: 'message_type', param_value: message_type }
        end

        # "version"が「1.3.0」でない場合、例外を発生する
        version = auth_hash.extra.raw_info["#{CLAIM}version"]
        raise InvalidLtiParam, { param_key: 'version', param_value: version } unless version == '1.3.0'

        # "roles"が空か無効値の場合、例外を発生する
        roles = auth_hash.extra.raw_info["#{CLAIM}roles"]
        raise InvalidLtiParam, { param_key: 'roles', param_value: roles } unless roles.present? && valid_roles?

        # "deployment_id"が空か無効値の場合、例外を発生する
        deployment_id = auth_hash.extra.raw_info["#{CLAIM}deployment_id"]
        unless deployment_id.present? && deployment_id.ascii_only? && deployment_id.length <= 255 &&
               deployment_id.start_with?('S_', 'B_', 'P_')
          raise InvalidLtiParam, { param_key: 'deployment_id', param_value: deployment_id }
        end

        # "context/id"が空の場合、例外を発生する
        context_id = auth_hash.extra.raw_info["#{CLAIM}context"]&.[]('id')
        unless context_id.present?
          raise InvalidLtiParam, { param_key: 'context.id', param_value: context_id }
        end

        # "resource_link/id"が空の場合、例外を発生する
        resource_link_id = auth_hash.extra.raw_info["#{CLAIM}resource_link"]&.[]('id')
        unless resource_link_id.present?
          raise InvalidLtiParam, { param_key: 'resource_link.id', param_value: resource_link_id }
        end

        # "target_link_uri"が空の場合、例外を発生する
        target_link_uri = auth_hash.extra.raw_info["#{CLAIM}target_link_uri"]
        unless target_link_uri.present?
          raise InvalidLtiParam, { param_key: 'target_link_uri', param_value: target_link_uri }
        end

        # "tool_platform"claimが存在し、かつ"tool_platform/guid"が空か無効値の場合、例外を発生する
        tool_platform = auth_hash.extra.raw_info["#{CLAIM}tool_platform"]
        if tool_platform.present?
          guid = tool_platform['guid']
          unless guid.present? && valid_uuid?(guid)
            raise InvalidLtiParam, { param_key: 'tool_platform.guid', param_value: guid }
          end
        end

        # "custom/grade"が空か無効値の場合、例外を発生する
        grade = auth_hash.extra.raw_info["#{CLAIM}custom"]&.[]('grade')
        unless grade.blank? || grade =~ /\A(P[1-6]|J[1-3]|H[1-3]|E[1-3])\z/i
          raise InvalidLtiParam, { param_key: 'custom.grade', param_value: grade }
        end
      end

      VOCAB_LIS = 'http://purl.imsglobal.org/vocab/lis/v2/'.freeze # 末尾スラッシュ有
      SYSTEM_ROLE = "#{VOCAB_LIS}system/person#".freeze # 末尾#有
      INSTITUTION_ROLE = "#{VOCAB_LIS}institution/person#".freeze # 末尾#有
      CONTEXT_ROLE = "#{VOCAB_LIS}membership#".freeze # 末尾#有
      CONTEXT_SUB_ROLE = "#{VOCAB_LIS}membership/".freeze # 末尾スラッシュ有
      ROLES = {
        # システムロール
        SYSTEM_ROLE => [
          'Administrator', 'None', # core
          'AccountAdmin', 'Creator', 'SysAdmin', 'SysSupport', 'User' # non-core
        ].freeze,
        # 教育機関ロール
        INSTITUTION_ROLE => [
          'Administrator', 'Faculty', 'Guest', 'None', 'Other', 'Staff', 'Student', # core
          'Alumni', 'Instructor', 'Learner', 'Member', 'Mentor', 'Observer', 'ProspectiveStudent' # non-core
        ].freeze,
        # コンテキストロール
        CONTEXT_ROLE => [
          'Administrator', 'ContentDeveloper', 'Instructor', 'Learner', 'Mentor', # core
          'Manager', 'Member', 'Officer' # non-core
        ].freeze,
        # コンテキストサブロール
        "#{CONTEXT_SUB_ROLE}Administrator#" => %w[
          Administrator Developer ExternalDeveloper ExternalSupport ExternalSystemAdministrator Support
          SystemAdministrator
        ].freeze,
        "#{CONTEXT_SUB_ROLE}ContentDeveloper#" => %w[
          ContentDeveloper ContentExpert ExternalContentExpert Librarian
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Instructor#" => %w[
          ExternalInstructor Grader GuestInstructor Lecturer PrimaryInstructor SecondaryInstructor TeachingAssistant
          TeachingAssistantGroup TeachingAssistantOffering TeachingAssistantSection TeachingAssistantSectionAssociation
          TeachingAssistantTemplate
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Learner#" => %w[
          ExternalLearner GuestLearner Instructor Learner NonCreditLearner
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Manager#" => %w[
          AreaManager CourseCoordinator ExternalObserver Manager Observer
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Member#" => %w[
          Member
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Mentor#" => %w[
          Advisor Auditor ExternalAdvisor ExternalAuditor ExternalLearningFacilitator ExternalMentor ExternalReviewer
          ExternalTutor LearningFacilitator Mentor Reviewer
        ].freeze,
        "#{CONTEXT_SUB_ROLE}Officer#" => %w[
          Chair Communications Secretary Treasurer Vice-Chair
        ].freeze
      }.freeze

      # "roles"が有効値かどうか判定する
      # @return 有効値: true/無効値: false
      def valid_roles?
        roles = auth_hash.extra.raw_info["#{CLAIM}roles"]
        # "roles"が空の場合、falseを返す
        return false if roles.blank?

        # "roles"にロール一覧に無い値が含まれる場合、falseを返す
        roles.each do |role|
          key = ROLES.keys.find { |k| role.start_with? k }
          return false if key.blank?

          return false unless ROLES[key].include?(role.delete_prefix(key))
        end
        # trueを返す
        true
      end

      # UUID(version4)として有効値かどうか判定する(有効値: true/無効値: false)
      # @param value 判定する値
      # @return 有効値: true/無効値: false
      def valid_uuid?(value)
        value =~ /\A[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[0-9a-f]{4}-[0-9a-f]{12}\z/i
      end

      # LTIパラメータ不正Error
      class InvalidLtiParam < StandardError
        attr_accessor :param_key, :param_value

        def initialize(data)
          super
          self.param_key = data[:param_key]
          self.param_value = data[:param_value]
        end

        def message
          "ログインに失敗しました。\nError: Invalid '#{param_key}' parameter"
        end
      end

      # 弊社独自領域ここから---
      ## 例外が発生した際に実行されるメソッド
      ## エラーをレスポンスする
      def render_500(e)
        logger.error "500 error #{e.message}"
        logger.error(e.backtrace.join("\n"))
        render_err e.message
      end
      # ---弊社独自領域ここまで
    end
  end
end
